src/ControlSystem/ControllerPND.cpp

```diff
@@ -0,0 +1,89 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "ControlSystem/ControllerPND.hpp"
+#include "ErrorHandling/Error.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+
+namespace Controllers {
+
+template <size_t deriv_order>
+DataVector PND<deriv_order>::operator()(
+    const DataVector& timescales,
+    const std::array<DataVector, deriv_order + 1>& q_and_derivs,
+    const double q_dt, const double deriv_dt) const noexcept {
+  // helper lambda for computing the binomial coefficients
+  auto binomial = [](size_t N, size_t k) {
+    return falling_factorial(N, k) / factorial(k);
+  };
+
+  // The coefficients a0_{k} are chosen such that the system is critically
+  // damped, which are given by: a0_{k} = (N choose k) timescales^{k-N}
+  std::array<DataVector, deriv_order> coefs0;
+  const DataVector inv_timescales = 1.0 / timescales;
+  DataVector tau = inv_timescales;
+  for (size_t i = deriv_order; i-- > 0;) {
+    gsl::at(coefs0, i) = binomial(deriv_order, i) * tau;
+    // add another power of timescales
+    tau = tau * inv_timescales;
+  }
+
+  // If an averager is used for q_and_derivs, there is an induced time offset,
+  // `t_offset`, due to the time-weighted averaging. Therefore, the q_and_derivs
+  // that we have in hand are at some time `t0`. However, we desire q_and_derivs
+  // at the current time `t = t0 + t_offset` to determine the appropriate
+  // control signal. We accomplish this by Taylor expanding `q(t0 + t_offset)`.
+  // The averager allows for averaging of q and its derivatives OR to not
+  // average q while still averaging the derivatives (the derivatives are always
+  // averaged in order to reduce noise due to numerical differentiation).
+  // When they are both averaged, the time offset will be identical for q and
+  // the derivatives (`q_dt = deriv_dt`). If a 'raw' q is used, then the time
+  // offset associated with q is zero (`q_dt=0`) and the derivative time offset
```

Comment:
I would move this documentation to the hpp so that it shows up in doxygen and also so that the parameter names  of `operator()` have context